docs/compilers/CSharp/Runtime Async Design.md

```diff
@@ -0,0 +1,456 @@
+#### Await a `T`-returning method
+
+```cs
+class C
+{
+    static T M<T>();
+}
+
+await C.M<Task>();
+```
+
+```il
+TODO: https://github.com/dotnet/runtime/issues/109632
+```
+
+#### Await a generic `T` `Task<T>`-returning method
+
+```cs
+class C
+{
+    static Task<T> M<T>();
+}
+
+await C.M<int>();
+```
+
+```il
+TODO: https://github.com/dotnet/runtime/issues/109632
+```
+
+#### Await a `Task`-returning delegate
+
+```cs
+delegate Task AsyncDelegate();
+
+class C
+{
+    static Task M();
+}
+
+AsyncDelegate d = C.M;
+await d;
+```
+
+```il
+TODO: https://github.com/dotnet/runtime/issues/109632
+```
+
+#### Await a `T`-returning delegate
+
+```cs
+
+class C
+{
+    static Task M();
+}
+
+Func<Task> d = C.M;
+await d;
+```
+
+```il
+TODO: https://github.com/dotnet/runtime/issues/109632
+```
+
+#### Awaiting in a `catch` block
+
+```cs
+class C
+{
+    static Task M();
+}
+
+try
+{
+    throw new Exception();
+}
+catch (Exception ex)
+{
+    await C.M();
+    throw;
+}
+```
+
+Translated C#:
+
+```cs
+int pendingCatch = 0;
+Exception pendingException;
+try
+{
+    throw new Exception();
+}
+catch (Exception e)
+{
+    pendingCatch = 1;
+    pendingException = e;
+}
+
+if (pendingCatch == 1)
+{
+    /* Runtime-Async Call */ C.M();
+    throw pendingException;
+}
+```
+
+```il
+{
+    .locals init (
+        [0] int32 pendingCatch,
+        [1] class [System.Runtime]System.Exception pendingException
+    )
+
+    .try
+    {
+        IL_0000: newobj instance void [System.Runtime]System.Exception::.ctor()
+        IL_0005: throw
+    }
+    catch [System.Runtime]System.Exception
+    {
+        IL_0006: stloc.1
+        IL_0007: ldc.i4.1
+        IL_0008: stloc.0
+        IL_0009: leave.s IL_000b
+    }
+
+    IL_000b: ldloc.0
+    IL_000c: ldc.i4.1
+    IL_000d: bne.un.s IL_0017
+
+    IL_000f: ldloc.1
+    IL_0010: call modreq(class [System.Runtime]System.Threading.Tasks.Task) void C::M()
+    IL_0015: pop
+    IL_0016: throw
+
+    IL_0017: ret
+}
+```
+
+#### Awaiting in a `finally` block
+
+```cs
+class C
+{
+    static Task M();
+}
+
+try
+{
+    throw new Exception();
+}
+finally
+{
+    await C.M();
+}
+```
+
+Translated C#:
+
+```cs
+Exception pendingException;
+try
+{
+    throw new Exception();
+}
+catch (Exception e)
+{
+    pendingException = e;
+}
+
+/* Runtime-Async Call */ C.M();
+
+if (pendingException != null)
+{
+    throw pendingException;
+}
+```
+
+```il
+{
+    .locals init (
+        [0] class [System.Runtime]System.Exception pendingException
+    )
+
+    .try
+    {
+        IL_0000: newobj instance void [System.Runtime]System.Exception::.ctor()
+        IL_0005: throw
+    }
+    catch [System.Runtime]System.Exception
+    {
+        IL_0006: stloc.0
+        IL_0007: leave.s IL_0009
+    }
+
+    IL_0009: call modreq(class [System.Runtime]System.Threading.Tasks.Task) void C::M()
+    IL_000e: pop
+    IL_000f: ldloc.0
+    IL_0010: brfalse.s IL_0014
+
+    IL_0012: ldloc.0
+    IL_0013: throw
+
+    IL_0014: ret
+}
+```
```

Comment:
How will we lower an `await` on a custom awaitable type with proper `GetAwaiter()`? I assume the same way described in section "Await local of concrete type `Task<T>`". Is that right?

We'll have to figure out what triggers the new special behavior for Task/Task<T>: are those special types or well-known types? I expect the runtime will need that we recognize only the special types (ie. Task/Task<T> from corlib)

Reply:
> How will we lower an await on a custom awaitable type with proper GetAwaiter()? I assume the same way described in section "Await local of concrete type Task<T>". Is that right?

Correct. I'll add that.